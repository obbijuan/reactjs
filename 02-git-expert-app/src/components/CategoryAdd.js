import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const CategoryAdd = ({setCategories} ) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => { 
        setInputValue(e.target.value)
    }

    const handleSubmit = (e) => { 
        e.preventDefault()

        if (inputValue.trim().length > 2) {
            // setCategories(categ => [inputValue, ...categ])
            setCategories([inputValue])
            setInputValue('')
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={ handleInputChange}
            />
        </form>
    )
}

CategoryAdd.propTypes = {
    setCategories: PropTypes.func.isRequired
}
