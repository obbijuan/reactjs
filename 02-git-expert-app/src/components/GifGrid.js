import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    // En la desectructuracion, se renombra data por images
    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>

            {loading && <p>Loading...</p>}

            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                        key={img.id}
                        {...img}
                        />
                        ))
                    }
            </div>
        </>
    )
}
