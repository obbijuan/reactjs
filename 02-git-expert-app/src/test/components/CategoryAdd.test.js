
import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import { CategoryAdd } from '../../components/CategoryAdd';


describe('Pruebas para el componente <CategoryAdd/>', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<CategoryAdd setCategories={ setCategories } />);
    
    // Se ejecuta para reinicializar
    beforeEach(() => { 
        jest.clearAllMocks();
        wrapper = shallow(<CategoryAdd setCategories={ setCategories } />);
    })

    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de cambiar el input', () => {
        
        const input = wrapper.find('input');
        const value = 'Agregando Categoria Test!'

        input.simulate('change', { target: { value } });
        // Obtiene el texto del parrafo, dentro del componente y lo compara
        // con el value que se simula para el input.
        expect(wrapper.find('p').text().trim()).toBe(value)

    })

    test('No debe de postear la informacion con submit', () => {
        wrapper.find('form').simulate('submit', { preventDefault: () => { } })
        // Valida que no se tiene que haber llamado
        expect(setCategories).not.toHaveBeenCalled()
    })

    test('debe de llamar el setCategories y limpiar el input', () => {
        
        const value = 'input test';

        // 1. Simular el inputChange
        wrapper.find('input').simulate('change', { target: { value } })
        
        // 2. Simular el submit
        wrapper.find('form').simulate('submit', { preventDefault(){ }})
        
        // 3. setCategories se debe de haber llamado
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith([value]);
        // expect(setCategories).toHaveBeenCalledWith( expect.any(Function) );
        
        // 4. El valor del input debe de estar ''
        expect(wrapper.find('input').prop('value')).toBe('');

        
    })
    
    
    
    
})
