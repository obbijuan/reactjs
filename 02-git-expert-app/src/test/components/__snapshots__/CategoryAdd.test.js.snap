// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para el componente <CategoryAdd/> debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
